src/act.movement.cpp

```diff
@@ -196,81 +197,99 @@ bool should_tch_see_chs_movement_message(struct char_data *tch, struct char_data
+          open_test_result = MIN(open_test_result - 2, 10);
+          strlcat(rbuf, "(defaulted: OT -2, cap 10!)", sizeof(rbuf));
+        } else {
+          // +4 means we defaulted to an attribute. Harsh penalty.
+          open_test_result = MIN(open_test_result - 4, 6);
+          strlcat(rbuf, "(defaulted: OT -4, cap 6!)", sizeof(rbuf));
+        }
       }
     }
 
-    int vision_penalty = get_vision_penalty(tch, in_room, rbuf, sizeof(rbuf));
-    snprintf(ENDOF(rbuf), sizeof(rbuf) - strlen(rbuf), ". TN is %d (OT) + %d (vis)", open_test_result, vision_penalty);
+    // Don't allow negative results.
+    open_test_result = MAX(open_test_result, 0);
 
-    int test_tn = open_test_result + vision_penalty;
+    // Vision penalties.
+    int vis_tn_modifier = calculate_vision_penalty(tch, ch);
+
+    // Other penalties.
+    int phys_tn_modifier = modify_target_rbuf_raw(tch, rbuf, sizeof(rbuf), vis_tn_modifier, FALSE);
 
     // House rule: Stealth/silence spells add a TN penalty to the spotter, up to 4.
-    {
-      int stealth_spell_tn = get_spell_affected_successes(ch, SPELL_STEALTH);
-      int silence_tn = in_room->silence[ROOM_NUM_SPELLS_OF_TYPE] > 0 ? in_room->silence[ROOM_HIGHEST_SPELL_FORCE] : 0;
-      int magic_tn_modifier = MIN(stealth_spell_tn + silence_tn, 4);
+    int stealth_spell_tn = get_spell_affected_successes(ch, SPELL_STEALTH);
+    int silence_tn = in_room->silence[ROOM_NUM_SPELLS_OF_TYPE] > 0 ? in_room->silence[ROOM_HIGHEST_SPELL_FORCE] : 0;
+    int magic_tn_modifier = MIN(stealth_spell_tn + silence_tn, 4);
 
-      if (magic_tn_modifier > 0) {
-        test_tn += magic_tn_modifier;
-        snprintf(ENDOF(rbuf), sizeof(rbuf) - strlen(rbuf), " + %d (magic)", magic_tn_modifier);
-      }
-    }
+    // Spirit Conceal power adds to the TN.
+    magic_tn_modifier += affected_by_power(ch, CONCEAL); 
+
+    snprintf(ENDOF(rbuf), sizeof(rbuf) - strlen(rbuf), ". TN is %d (OT) + %d (vision) + %d (other phys) + %d (other magic)",
+                                                      open_test_result, vis_tn_modifier, phys_tn_modifier, magic_tn_modifier);
 
     // Roll the perception test.
+    int test_tn = open_test_result + vis_tn_modifier + phys_tn_modifier + magic_tn_modifier;
     int perception_result = success_test(GET_INT(tch) + GET_POWER(tch, ADEPT_IMPROVED_PERCEPT), test_tn);
     snprintf(ENDOF(rbuf), sizeof(rbuf) - strlen(rbuf), "Result: %d hits.", perception_result);
     if (GET_INVIS_LEV(tch) <= GET_LEVEL(ch))
       act(rbuf, FALSE, tch, 0, 0, TO_STAFF_ROLLS);
 
     bool spotted_movement = perception_result > 0;
 
-    // We rolled at least one hit above our open test TN? NOW we check if we can see them.
-    if (spotted_movement && CAN_SEE(tch, ch)) {
+    // Failed to notice them.
+    if (!spotted_movement) {
+      return FALSE;
+    }
+
+    // They were sneaking but we noticed.
+    if (IS_AFFECTED(ch, AFF_SNEAK) && spotted_movement) {
```

Comment:
Let's make this check a `spotted_movement && (IS_AFFECTED(ch, AFF_SNEAK) || !CAN_SEE(ch))` since hearing an invis moving is also sus.

Reply:
The "can't see but we noticed case" looks like its handled from the result of the `should_tch_see_chs_movement_message` call on line 440, which in turn calls `process_spotted_invis`. Should that be moved from that part of the code to inside `should_tch_see_chs_movement_message` itself?